subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/services/DefaultEnvironmentChangeTracker.kt

```diff
@@ -105,19 +119,47 @@ class DefaultEnvironmentChangeTracker(private val problemFactory: ProblemFactory
         }
 
         fun isSystemPropertyMutated(key: String): Boolean {
-            return systemPropertiesCleared || mutatedSystemProperties.containsKey(key)
+            return systemPropertiesCleared || mutatedSystemProperties[key] is SystemPropertyMutate
+        }
+
+        fun isSystemPropertyLoaded(key: String): Boolean {
+            return mutatedSystemProperties[key] is SystemPropertyLoad
+        }
+
+        fun getLoadedPropertyOldValue(key: String): Any? {
+            return (mutatedSystemProperties[key] as? SystemPropertyLoad)?.oldValue
         }
 
         fun getCachedState(): CachedEnvironmentState {
             return CachedEnvironmentState(
                 cleared = systemPropertiesCleared,
-                updates = mutatedSystemProperties.values.filterIsInstance<SystemPropertySet>(),
+                updates = mutatedSystemProperties.values
+                    .filterIsInstance<SystemPropertySet>()
+                    .filter { it !is SystemPropertyOverride },
                 removals = mutatedSystemProperties.values.filterIsInstance<SystemPropertyRemove>()
             )
         }
 
+        fun systemPropertyOverridden(key: Any, value: Any?) {
+            mutatedSystemProperties[key] = SystemPropertyOverride(key, value)
```

Comment:
Can we avoid having `SystemPropertyOverride` at all, and just remove `key` from `mutatedSystemProperties` here? Here's my understanding. You have `SystemPropertyOverride` binding after root build properties evaluation, when you have `-Dfoo`. Its presence doesn't affect reads of the variable. It seems to matter is `systemPropertyLoaded`, when you're redefining the property for a brief moment. However, you'll get the value from `System.getProperties` as an `oldValue` here anyway. And the next step would be calling `systemPropertyOverridden` again. True, it is important for us not to have `key` in the stored state, but we don't have to actually keep track of the overridden property for that, we can just discard the mapping.